packages/expect/__typetests__/expect.test.ts

```diff
@@ -79,3 +86,128 @@ expectType<void>(
     bananas: expect.not.toBeWithinRange(11, 20),
   }),
 );
+
+// MatcherFunction
+
+expectError(() => {
+  const actualMustBeUnknown: MatcherFunction = (actual: string) => {
+    return {
+      message: () => `result: ${actual}`,
+      pass: actual === 'result',
+    };
+  };
+});
+
+expectError(() => {
+  const lacksMessage: MatcherFunction = (actual: unknown) => {
+    return {
+      pass: actual === 'result',
+    };
+  };
+});
+
+expectError(() => {
+  const lacksPass: MatcherFunction = (actual: unknown) => {
+    return {
+      message: () => `result: ${actual}`,
+    };
+  };
+});
+
+type ToBeWithinRange = (
+  this: MatcherState,
+  actual: unknown,
+  floor: number,
+  ceiling: number,
+) => any;
+
+const toBeWithinRange: MatcherFunction<[floor: number, ceiling: number]> = (
+  actual: unknown,
+  floor: unknown,
+  ceiling: unknown,
+) => {
+  return {
+    message: () => `actual ${actual}; range ${floor}-${ceiling}`,
+    pass: true,
+  };
+};
+
+expectAssignable<ToBeWithinRange>(toBeWithinRange);
+
+type AllowOmittingExpected = (this: MatcherState, actual: unknown) => any;
+
+const allowOmittingExpected: MatcherFunction<[]> = (
```

Comment:
The `<[]>` is unnecessary here. I just wanted to show why the default of `type MatcherFunction<Expected extends Array<unknown> = []> =` is `[]`. This second generic type argument is used just to infer types of args in `expect.extend()` (not yet implemented). The current default is – matcher takes zero expected args. If a matcher takes any expected args, these must be declared explicitly. If default is not `[]` zero expected args taking matchers have to declare `<[]>`. At least this is the limitation I couldn’t overcome.

Reply:
That sounds reasonable to me (I think 😀)

Reply:
Great. So I will remove it. That was just a demo.